src/web/vaev-driver/fetcher.h

```diff
@@ -2,18 +2,216 @@
 
 #include <karm-gc/heap.h>
 #include <karm-mime/url.h>
+#include <karm-net/http/http.h>
+#include <karm-sys/chan.h>
+#include <karm-sys/file.h>
 #include <vaev-dom/document.h>
 #include <vaev-style/stylesheet.h>
 
 namespace Vaev::Driver {
 
-Res<Style::StyleSheet> fetchStylesheet(Mime::Url url, Style::Origin origin = Style::Origin::AUTHOR);
+struct ChunkedTransfer : public Io::Writer {
+    bool isDone = false;
+    Io::Writer& _out;
 
-void fetchStylesheets(Gc::Ref<Dom::Node> node, Style::StyleBook& sb);
+    virtual Res<usize> _write(Bytes bytes) = 0;
 
-Res<Gc::Ref<Dom::Document>> fetchDocument(Gc::Heap& heap, Mime::Url const& url);
+    Res<usize> write(Bytes bytes) override {
+        if (isDone)
+            return Error::brokenPipe("Cannot write to transfer after it is done.");
+        return _write(bytes);
+    }
 
-Res<Gc::Ref<Dom::Document>> loadDocument(Gc::Heap& heap, Mime::Url const& url, Mime::Mime const& mime, Io::Reader& reader);
+    virtual Res<usize> _done() = 0;
+
+    Res<usize> done() {
+        auto written = try$(_done());
+        isDone = true;
+        return Ok(written);
+    }
+
+    ChunkedTransfer(Io::Writer& out) : _out(out){};
+};
+
+struct FileChunkTransfer : public ChunkedTransfer {
+
+    FileChunkTransfer(Io::Writer& out) : ChunkedTransfer(out) {}
+
+    Res<usize> _write(Bytes bytes) override {
+        return _out.write(bytes);
+    }
+
+    Res<usize> _done() override {
+        return Ok(0);
+    }
+};
+
+struct Fetcher {
+
+    Opt<Rc<Fetcher>> fallback;
+
+    Fetcher(Opt<Rc<Fetcher>> fallback = NONE) : fallback(fallback) {}
+
+    virtual Res<String> fetch(Mime::Url const& url) = 0;
+
+    virtual Res<Rc<ChunkedTransfer>> transfer(Mime::Url const& url) = 0;
+};
+
+struct FileFetcher : public Fetcher {
+
+    virtual ~FileFetcher() = default;
+
+    Mime::Url const STDIN_URL = "about:stdin"_url;
+    Mime::Url const STDOUT_URL = "about:stdout"_url;
+
+    Res<String> fetch(Mime::Url const& url) override {
+        if (url == STDIN_URL)
+            return Io::readAllUtf8(Sys::in());
+
+        auto file = try$(Sys::File::open(url));
+        return Io::readAllUtf8(file);
+    }
+
+    virtual Res<Rc<ChunkedTransfer>> transfer(Mime::Url const& url) override {
+        if (url == STDOUT_URL)
+            return Ok(makeRc<FileChunkTransfer>(Sys::out()));
+
+        Sys::FileWriter fileWriter{try$(Sys::File::create(url))};
+        return Ok(makeRc<FileChunkTransfer>(fileWriter));
```

Comment:
use-after-free here

Reply:
yep, thx, forgot this

Reply:
my plan is to fix this putting a `Rc` but it depends on the resolution of the first comment